### ISSUE TITLE
Krav Maga Block Counter irregularities

### ISSUE THREAD

**Issue by JinJanODWan** [CleverRaven/Cataclysm-DDA#46671]
# Describe the bug
Krav Maga **Block Counter** does not proc with style weapons like the USP and combat knife.

# Steps To Reproduce
1. Create a player with an USP, combat knife, and tiger claws
2. Learn Krav Maga
3. Set Unarmed skill to 4+
4. Find something that hits back
3. Try to block using above weapons

# Expected behavior
Player occasionally blocks incoming damage using the USP or combat knife, if I'm reading [this](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/json/techniques.json#L960) correctly.

# Screenshots
Log shows *other* Krav Maga techniques being used, but no Block Counters proc
![kravmagalog](https://user-images.githubusercontent.com/44056502/104145568-2b518d80-53d0-11eb-87fa-738feb56776f.png)

# Versions and configuration
 - OS: Windows 10
 - Game Version: 0.E-8852-gf360c7f
 - Graphics version: Tiles
 - Ingame language: English
 - Mods loaded: dda, no_npc_food

# Additional context
Tiger Claws and other unarmed seem to proc arm- as well as leg blocks as expected.

**Comment by int-ua** [comment]
What's `proc` in these contexts?

**Comment by Zireael07** [comment]
Does not 'work' or 'run'

**Comment by JinJanODWan** [comment]
Thanks for having my back, Zireael! 
Sorry; I assumed _proc_ is more widely used than it is. I believe it is an acronym for _programmed random occurrence_ (n) or _programmatically randomly occur_ (v).
Just to be clear: the blocks seemingly never occur with the USP or Combat Knife, but definitely does occur on the same character when unarmed or using weapons with the UNARMED flag, like the Tiger Claws.

**Comment by hexagonrecursion** [comment]
There are two types of blocks:
* Unarmed blocks (arm blocks and leg blocks) - they do not trigger when wielding a weapon unless it has the UNARMED flag
* Blocking with a weapon - some weapons have techniques: WBLOCK_1, WBLOCK_2 or WBLOCK_3

The combat knife is not unarmed preventing arm and leg blocks. It also does not have a block technique preventing weapon blocks.

Example of the counter activating on a weapon block:

![block](https://user-images.githubusercontent.com/52621858/129186359-bfdacc43-e768-4e1c-af9d-e4e2180de1df.png)

I think this issue should be closed

**Comment by JinJanODWan** [comment]
_Thanks for your time as well as the clarification! I know the norm is to get a strong gun and keep your distance, but I would love to see more from melee and I think this has highlighted a current shortfall of the system._

My problem with the way the blocking system currently works is twofold:

1. One's martial arts seemingly becomes less effective in combat when armed, which seems counter-intuitive, IMHO
2. Some weapons (e.g. katana, cudgel, and quarterstaff, off the top of my head) seemingly allow very effective parries even to a character with 0 melee and 0 unarmed

- I can understand how using martial arts' arm blocking techniques while wielding a firearm is asking for trouble, but from the same practical perspective, I struggle to understand how you wouldn't be able to use Krav Maga's (and some other styles that have a heavy focus on using weapons e.g. Eskrima and Silat, to name but two) arm blocks with style-specific weapons in your hand (e.g. Krav Maga + Combat Knife), especially if said style has very practical application of said weapon and that same character can block the same incoming attack when unarmed. Similarly, does it really make sense that you suddenly can't use your legs to block that same attack you would have been able to if you were unarmed just because you happen to be wielding a knife or pistol right at this very moment?

### Potential solutions I considered
My first thought was: "Should these knives and other style weapons then have WBLOCK_x flag added?", but that doesn't really make sense if they're wielded by a character who doesn't have the relevant martial arts training to use it effectively.

Maybe one should add the UNARMED flag to these weapons to better represent what they might perform like in the hands of a skilled fighter, especially as the actual blocking will be done by the arm (or leg) rather than the knife. However, I'm not sure that this is the best solution as it might create chaos elsewhere and doesn't account for melee skill as far as I'm aware.

### What I would like to see:
I know this will be tedious, but maybe weapon blocking as a whole needs a rework where:

1. arm and leg blocks still work when the character has relevant style and skills and are wielding style-specific weapons
2. the WBLOCK_x techniques only proc when a character has high enough Melee skill when using styles 

Yes, this would make a cudgel, quarterstaff, and katana (to name a few) less useful from the get-go, but I would argue that defense should scale as offense does. Just as there are inferior and superior ways to strike your opponent, there are inferior and superior ways to attempt a block. It would both be more realistic as well as adding depth to the gameplay. 

This would equate to a martial artists (or skilled Brawler) being able to block with their arms or legs if they have the relevant skill irrespective of whether they are unarmed or armed with a style weapon. Weapon specific counters as seen in your screenshot can and should still occur as normal, but using a more defensive weapon with your style will equate to overall better defense.

### Potential Implementation
Melee blocks would unlock in a scaling way the same way as the the unarmed arm and leg blocks do in the current game:

- WBLOCK_1 to proc on any weapons with WBLOCK_1 and higher only once a basic melee skill, e.g. Melee 2, is reached
- WBLOCK_2 requires more skill and, as such, only unlocks once a higher skill, e.g. Melee 4 is reached
- WBLOCK_3 is only fully utilized later, e.g. Melee 5
 
Many weapons could realistically have WBLOCK_1 added to potentially offset the nerf that is introduced somewhat. I don't personally see why one can't use, for example, an arming sword or even a kirpan to at least partially block an attack. 

However, the above proposed solution still doesn't account for Unarmed skill level and martial arts, which is not ideal.

**Comment by Hymore246** [comment]
I don't want to go into exhaustive detail about the blocking formula since I do not have time at the moment but I will address a few points.

* In order to block an attack with an item, it must have a WBLOCK tag. The higher the tag number, the more "effective" the item is at blocking, i.e. reducing damage.
* The WBLOCK tag represent how effective an item is to block attacks with. It has nothing to do with skill. A person with Broadsword can block an attack with it. Note that I didn't say block _effectively_. An unskilled person would still probably take the brunt of the damage but they would still try to reflectively interpose the Broadsword between themselves and an attack. Combat styles require unarmed skill to block because the human body isn't quite as study as a metal object and the average person doesn't know how to position themselves to stop an attack period. Reflexively raising your hand in response a punch isn't going to do much. So having requirements attached to the blocking does nothing but hurt the player. 
* Each of the WBLOCK tags are based on the size of the weapon as far as I can tell. There isn't any documentation but WBLOCK_1 goes to smaller weapons such as knives, WBLOCK_2 is for bigger weapons such as the longsword and zweihander, WBLOCK_3 is exclusively for shields and when I say "shields" I mean a literal medieval shields. No WBLOCK_3 items exist in the base game at the moment.
* Melee skill is already factored into the block formula. So a person with Melee 0 is less effective than a person with Melee 5.
* I agree more items could have the blocking tags. Which ones are subjective. I could argue that I could block an attack with a 100L drum. Not effectively but the attack would have to go through the drum first. 
* Giving guns WBLOCK isn't overpowered. In fact #49445 wants to do this with rifles. The danger when blocking with a gun is that they are not made for stopping attacks. They are not sturdy and you risk damaging the weapon when you block. That is what would make the change balanced. Also, being point blank with an enemy is not the best place to start aiming your gun.
* Lastly, your new suggestion falls outside the scope of the issue you made. You might want to make a new issue about WBLOCK and blocking if you still want to continue with your suggestion.

Lastly, this issue can be closed once #51949 is completed. Dodge and Block counters are no longer in the game after the update is complete.

**Comment by Night-Pryanik** [comment]
Closing as stale, since stalebot can't do this by itself.
